tests/test_accelerator.py

```diff
@@ -703,77 +703,57 @@ def test_save_model_with_stateful_dataloader(self, use_safetensors, tied_weights
-
-            original_linear1 = prepared_model.linear1.weight.clone()
-            original_batchnorm = prepared_model.batchnorm.weight.clone()
-            original_linear2 = prepared_model.linear2.weight.clone()
-
-            # Load the model and state dict
-            load_checkpoint_in_model(model, tmpdirname)
-            stateful_train_dl, _ = create_dataloaders_for_test(num_workers=num_workers)
-            prepared_stateful_train_dl = accelerator.prepare_data_loader(stateful_train_dl)
-            prepared_stateful_train_dl.load_state_dict(state_dict)
+                if step == num_batches_to_skip - 1:
+                    # Save the state once we've gone through a few batches
+                    accelerator.save_state(f"{tmpdirname}/state", safe_serialization=use_safetensors)
+                    # When breaking out without fully going through the iterator, must call end() to unregister this iterator from gradient state.
+                    # TODO: Maybe this could be done automatically?
+                    # prepared_train_dl.end()
+                    # break
+                if step >= num_batches_to_skip:
+                    untrained_batches.append(batch)
+
+            not_skipped_batches = accelerator.gather(untrained_batches)
+            # We then unwrap the trained model
+            unwrapped_model = accelerator.unwrap_model(prepared_model)
+
+            original_linear1 = unwrapped_model.linear1.weight.clone()
+            original_batchnorm = unwrapped_model.batchnorm.weight.clone()
+            original_linear2 = unwrapped_model.linear2.weight.clone()
+
+            # Resume the state
+            accelerator.load_state(f"{tmpdirname}/state")
 
             # Train this to the end of the DataLoader
             batches_seen_with_loaded_dl = 0
-            for batch in prepared_stateful_train_dl:
+            for batch in prepared_train_dl:
                 x, y = batch
-                x.to(accelerator.device)
-                y.to(accelerator.device)
-                with accelerator.accumulate(prepared_model):
-                    outputs = prepared_model(x)
-                    loss = torch.nn.functional.mse_loss(outputs, y)
-                    accelerator.backward(loss)
-                    prepared_optimizer.step()
-                    prepared_scheduler.step()
-                    prepared_optimizer.zero_grad()
+                outputs = prepared_model(x)
+                loss = torch.nn.functional.mse_loss(outputs, y)
+                accelerator.backward(loss)
+                prepared_optimizer.step()
+                prepared_scheduler.step()
+                prepared_optimizer.zero_grad()
                 batches_seen_with_loaded_dl += 1
 
-            new_linear1 = prepared_model.linear1.weight
-            new_batchnorm = prepared_model.batchnorm.weight
-            new_linear2 = prepared_model.linear2.weight
+            unwrapped_model_2 = accelerator.unwrap_model(prepared_model)
```

Comment:
during any distributed process, we need to unwrap the model first before we can toy with the layers